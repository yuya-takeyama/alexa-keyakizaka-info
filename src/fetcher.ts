import axios from 'axios';
import { JSDOM } from 'jsdom';
import { first, map, tail, trim } from 'lodash';
import * as moment from 'moment';
import * as striptags from 'striptags';

const BASE_URL = 'http://www.keyakizaka46.com';

const toURL = (path: string): string => {
  return `${BASE_URL}${path}`;
};

export interface Schedule {
  title?: string;
  time?: string;
  genre?: string;
  description?: string;
}

const normalize = (str: string | undefined | null): string | undefined => {
  return str ? trim(striptags(str)) || undefined : undefined;
};

const parseScheduleElement = (element: Element): Schedule | undefined => {
  const titleDescriptionElement = element.querySelector(
    '.box-detail_txt :last-child',
  );
  const genreElement = element.querySelector('.box-detail_genre');
  const timeElement = element.querySelector('.box-detail_txt .time');

  if (titleDescriptionElement) {
    const titleDescriptions = trim(
      striptags(titleDescriptionElement.innerHTML),
    ).split(/\n/);
    const title = first(titleDescriptions);
    const description = normalize(tail(titleDescriptions).join('\n'));

    return {
      title,
      genre: genreElement ? normalize(genreElement.innerHTML) : undefined,
      time: timeElement ? normalize(timeElement.innerHTML) : undefined,
      description,
    };
  } else {
    return undefined;
  }
};

const today = (): moment.Moment => moment.utc({ hour: 0 });

export const fetchSchedules = async (
  date: moment.Moment = today(),
): Promise<Schedule[]> => {
  return new Promise<Schedule[]>(async (resolve, reject) => {
    try {
      const res = await axios.get(
        toURL(`/s/k46o/media/list?dy=${date.format('YYYYMMDD')}`),
      );
      const scheduleElements = new JSDOM(
        res.data,
      ).window.document.querySelectorAll(
        '#schedule .box-schedule_inner .box-detail',
      );
      const schedules = map(scheduleElements, element =>
        parseScheduleElement(element),
      ).filter(schedule => schedule) as Schedule[];

      resolve(schedules);
    } catch (err) {
      reject(err);
    }
  });
};
