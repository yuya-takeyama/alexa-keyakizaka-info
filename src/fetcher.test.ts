import { parseTime } from "./fetcher";

describe('#parseTime', () => {
  describe('when the time is undefined', () => {
    it('returns from and to as undefined', () => {
      expect(parseTime(undefined)).toEqual({ from: undefined, to: undefined });
    });
  });

  describe('when the time has both of from and to', () => {
    it('returns from and to', () => {
      expect(parseTime('19:00～20:00')).toEqual({ from: '19:00', to: '20:00' });
    });
  });

  describe('when the time has from only', () => {
    it('returns from only', () => {
      expect(parseTime('19:00～')).toEqual({ from: '19:00', to: undefined });
    });
  });

  describe('when the time has to only', () => {
    it('returns to only', () => {
      expect(parseTime('～20:00')).toEqual({ from: undefined, to: '20:00' });
    });
  });
});
