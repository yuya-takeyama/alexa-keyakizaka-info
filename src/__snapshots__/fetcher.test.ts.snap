// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#toValidJSON converts from JavaScript to JSON 1`] = `
"
[
{
    \\"title\\": \\"テレビ東京「木ドラ25」『\\\\\\"Re\\\\\\":Mind』\\",
    \\"start\\": \\"2017-11-02\\",
    \\"className\\": \\"media\\",
    \\"description\\":\\"テレビ東京「木ドラ25」『\\\\\\"Re\\\\\\":Mind』\\"
  },{
    \\"title\\": \\"「MEN'S NON-NO」12月号\\",
    \\"start\\": \\"2017-11-10\\",
    \\"className\\": \\"media\\",
    \\"description\\":\\"「MEN'S NON-NO」12月号\\"
  },{
    \\"title\\": \\"「FINEBOYS」12月号\\",
    \\"start\\": \\"2017-11-10\\",
    \\"className\\": \\"media\\",
    \\"description\\":\\"「FINEBOYS」12月号\\"
  }
]
    "
`;
